File: packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/exporters/CompressionImageExporter.java
package expo.modules.imagepicker.exporters;

import android.graphics.Bitmap;
import android.net.Uri;

import org.apache.commons.io.FilenameUtils;
import org.jetbrains.annotations.NotNull;
import org.unimodules.interfaces.imageloader.ImageLoader;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public class CompressionImageExporter implements ImageExporter {
  private ImageLoader mImageLoader;
  private int mQuality;
  private boolean mBase64;

  public CompressionImageExporter(@NonNull final ImageLoader imageLoader, int quality, boolean base64) {
    mImageLoader = imageLoader;
    mQuality = quality;
    mBase64 = base64;
  }

  @Override
  public void export(@NotNull Uri source, @NotNull File output, @NotNull ImageExporterListener exporterListener) {
    mImageLoader.loadImageForManipulationFromURL(source.toString(), new ImageLoader.ResultListener() {

      @Override
      public void onSuccess(@NonNull Bitmap bitmap) {
        int width = bitmap.getWidth();
        int height = bitmap.getHeight();
        ByteArrayOutputStream out = mBase64 ? new ByteArrayOutputStream() : null;

        try {
          Bitmap.CompressFormat compressFormat = Bitmap.CompressFormat.JPEG;
          if (FilenameUtils.getExtension(output.getPath()).contains("png")) {
            compressFormat = Bitmap.CompressFormat.PNG;
          }

          saveBitmap(bitmap, compressFormat, output, out);
          exporterListener.onExportResult(out, width, height);
        } catch (IOException e) {
          exporterListener.onFailure(e);
        }

        if (out != null) {
          try {
            out.close();
          } catch (IOException e) {
            // We have already resolve promise.
            e.printStackTrace();
          }
        }
      }

      @Override
      public void onFailure(@Nullable Throwable cause) {
        exporterListener.onFailure(cause);
      }
    });
  }

  /**
   * Compress and save the {@code bitmap} to {@code file}, optionally saving it in {@code out} if
   * base64 is requested.
   *
   * @param bitmap         bitmap to be saved
   * @param compressFormat compression format to save the image in
   * @param output         file to save the image to
   * @param out            if not null, the stream to save the image to
   */
  private void saveBitmap(Bitmap bitmap, Bitmap.CompressFormat compressFormat, File output,
                          ByteArrayOutputStream out) throws IOException {
    writeImage(bitmap, output, compressFormat);

    if (mBase64) {
      bitmap.compress(Bitmap.CompressFormat.JPEG, mQuality, out);
    }
  }

  private void writeImage(Bitmap image, File output, Bitmap.CompressFormat compressFormat) throws IOException {
    try (FileOutputStream out = new FileOutputStream(output)) {
      image.compress(compressFormat, mQuality, out);
    }
  }

}

File: packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/exporters/CompressionImageExporter.kt
package expo.modules.imagepicker.exporters

import android.content.ContentResolver
import android.graphics.Bitmap
import android.net.Uri
import androidx.annotation.FloatRange
import androidx.core.net.toFile
import expo.modules.imagepicker.FailedToReadFileException
import expo.modules.imagepicker.FailedToWriteFileException
import expo.modules.imagepicker.MissingModuleException
import expo.modules.imagepicker.copyExifData
import expo.modules.imagepicker.toBitmapCompressFormat
import expo.modules.kotlin.providers.AppContextProvider
import kotlinx.coroutines.runInterruptible
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.util.concurrent.ExecutionException

class CompressionImageExporter(
  private val appContextProvider: AppContextProvider,
  @FloatRange(from = 0.0, to = 1.0)
  quality: Double
) : ImageExporter {
  private val compressQuality = (quality * 100).toInt()

  override suspend fun exportAsync(
    source: Uri,
    output: File,
    contentResolver: ContentResolver
  ): ImageExportResult {
    val bitmap = readBitmap(source)
    val compressFormat = output.toBitmapCompressFormat()
    writeImage(bitmap, output, compressFormat)
    copyExifData(source, output, contentResolver)

    return object : ImageExportResult(
      bitmap.width,
      bitmap.height,
      output
    ) {
      override suspend fun data(contentResolver: ContentResolver): ByteArrayOutputStream {
        val outputStream = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, compressQuality, outputStream)
        return outputStream
      }
    }
  }

  private suspend fun readBitmap(source: Uri): Bitmap = runInterruptible {
    val loaderResult = appContextProvider.appContext.imageLoader
      ?.loadImageForManipulationFromURL(source.toString())
      ?: throw MissingModuleException("ImageLoader")

    try {
      loaderResult.get()
    } catch (cause: ExecutionException) {
      throw FailedToReadFileException(source.toFile(), cause)
    }
  }

  /**
   * Compress and save the `bitmap` to `file`
   * @throws [IOException]
   */
  private suspend fun writeImage(
    bitmap: Bitmap,
    output: File,
    compressFormat: Bitmap.CompressFormat
  ) = runInterruptible {
    try {
      FileOutputStream(output).use { out -> bitmap.compress(compressFormat, compressQuality, out) }
    } catch (cause: FileNotFoundException) {
      throw FailedToWriteFileException(output, cause)
    }
  }
}
